### PR TITLE
Added inductor_prims.unsafe_index_put_

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112106
* #112102
* #112101
* #112093
* #111990

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_perf.py [hunk 1]
```diff
@@ -596,12 +596,18 @@ def f2(a):
 class InplacingTests(TestCase):
     def test_inplace_scatter(self):
         def f(a, b):
-            a = a.cos()
-            a[b] = 1
```
Comment (lezcano) [inline test/inductor/test_perf.py:600]: Do not remove the previous test. The advanced indexing path is different to directly calling `index_put_`

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -665,7 +665,7 @@ def can_replace(node, mutated_arg):
             if len(shared_view_nodes) > 2:  # Arg aliases another node other than copy_
                 return False
 
-            # Check for any uses other than current node and copy_ epilogue
+            # # Check for any uses other than current node and copy_ epilogue
```
Comment (lezcano) [inline torch/_inductor/fx_passes/post_grad.py:668]: comment the comment? lol

File: torch/_inductor/fx_passes/post_grad.py [hunk 3]
```diff
@@ -682,12 +682,15 @@ def can_replace(node, mutated_arg):
 
     inplaceable_ops = {
         aten.index_put.default: InplaceableOp(aten.index_put_.default, 0),
+        aten._unsafe_index_put.default: InplaceableOp(
+            aten._unsafe_index_put_.default, 0
+        ),
```
Comment (lezcano) [inline torch/_inductor/fx_passes/post_grad.py:687]: Just a side comment, but do we have any guarantees that the rest of the inductor stack knows how to handle in-place ops at all?

**Review by lezcano — COMMENTED** [review]
Ok, so this one is a bit funny. This one is not meant to be used externally (it doesn't have a derivative, a meta function or anything) so it's just internal to dynamo. Even more, it is internal at an IR level as it is not meant to be used by any backend other than inductor (it shouldn't appear in decompositions). Could you leave a comment in native_functions.yaml explaining a bit the reason of being of this function and these limitations?

**Comment by Chillee** [comment]
Actually, I might just make it an inductor internal op 🤔

File: torch/_inductor/fx_passes/post_grad.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/fx_passes/post_grad.py:687]: Yes, the inductor stack post fx passes supports in place operations. However, the graph passes don’t. Check fx_oasses/readme for the  invariants

**Review by lezcano — APPROVED** [review]
So much more self-contained. Nice!

**Comment by Chillee** [comment]
After reverting a previous commit/unreverting it, this PR somehow got merged into https://github.com/pytorch/pytorch/pull/112102/files#diff-32eb31ed8659f8f9edf7a939460b83419db0ea45ec85dd5fc6e67e70cb426ebf.

**Pull request closed without merging**